### DISCUSSION TITLE
Getting the source branch on a `pull_request` event and the pushed branch on a `push` event

### DISCUSSION THREAD

**Discussion by elongl** [community/community#30794] [Actions] [answered]
I want a way to get the **branch name** (not `<NUM>/pull`, or `<NUM>/merge`, etc.) of the source branch on a PR and the pushed branch when pushing.  

Initially I thought that `github.ref_name` would give me that but instead it returned `270/merge` and not the PR's branch name.
I then started using `github.head_ref`, but it only applies to PRs and not to pushed branches.

What's an expression that would evaluate to `github.head_ref` on PR and to the pushed branch on a `push` event?

Thanks.

**Comment by Yuri05** [comment]
`github.event.pull_request.head.ref` ?

**Comment by ViacheslavKudinov** [comment]
Hi,
maybe
`github.base_ref `- The base_ref or target branch of the pull request in a workflow run. This property is only available when the event that triggers a workflow run is either pull_request or pull_request_target

You can check `github` context to see what else is available https://docs.github.com/en/actions/learn-github-actions/contexts#github-context.

**Comment by elongl** [comment, marked as answer]
I ended up using `BRANCH_NAME: ${{ github.head_ref || github.ref_name }}` but I'm not sure if this is the best way, works for now.

**Reply by alexjanckilagc** [reply]
This does not work and is an empty value on PR merge.

**Reply by krebemm** [reply]
is there any fix for this? push request would be a lot more useful with github.head_ref variable :)

**Reply by aghand0ur** [reply]
any update on this matter?

**Reply by McBlankenburg** [reply]
i need it too, i want scrap branch name to set labels in pull request template

**Comment by syedammadiqbal** [comment]
To get source branch Name : ${{github.ref_name}} 
For Target branch Name :  ${{github.event.pull_request.base.ref}}

**Comment by recroviral** [comment]
I have tired to retrieve Branch name from the merge commit message on a push event and it's working for me. 

```
# Get the Branch Name
MERGE_COMMIT_MESSAGE=$(git log --format=%B -n 1 "${GITHUB_SHA}")
BRANCH_NAME=$(echo "$MERGE_COMMIT_MESSAGE" | grep -oP '(?<=from ).*' | awk -F/ '{print $NF}')
echo "Branch Name: $BRANCH_NAME"
```

**Reply by rkochar** [reply]
> I have tired to retrieve Branch name from the merge commit message on a push event and it's working for me.
> 
> ```
> # Get the Branch Name
> MERGE_COMMIT_MESSAGE=$(git log --format=%B -n 1 "${GITHUB_SHA}")
> BRANCH_NAME=$(echo "$MERGE_COMMIT_MESSAGE" | grep -oP '(?<=from ).*' | awk -F/ '{print $NF}')
> echo "Branch Name: $BRANCH_NAME"
> ```

How do you set the value in `${GITHUB_SHA}`?

**Reply by recroviral** [reply]
"${GITHUB_SHA}": the SHA hash of the commit provided by GitHub Actions as the SHA of the commit that triggered the workflow.

You can directly use GITHUB_SHA without needing to set it yourself.

**Reply by rkochar** [reply]
> "${GITHUB_SHA}": the SHA hash of the commit provided by GitHub Actions as the SHA of the commit that triggered the workflow.
> 
> You can directly use GITHUB_SHA without needing to set it yourself.

Thank you for your response. I think merge queues break this workaround.

**Reply by rhombl4** [reply]
Keep in mind that GITHUB_SHA is the sha of merge-commit, not the commit in head branch. You won't find this sha in a repository after referencing it somewhere; in the case of `"${{ github.event_name }}" == "pull_request"`.

**Comment by avelinoappfire** [comment]
i was able to do it with this solution from stack overflow: https://stackoverflow.com/a/58035262
hope it helps

**Reply by avelinoappfire** [reply]
oh but no when the branch is merged after a PR, has someone managed to get the source branch name after it has been merged?

**Reply by pi3k14** [reply]
`gh api repos/${{ github.repository }}/commits/${{ github.sha }}/pulls --jq '.[0].head.ref'` is the proper spell I think
